01-messaging.md

```diff
@@ -82,6 +83,104 @@ however, adding a 6-byte padding after the type field was considered
 wasteful: alignment may be achieved by decrypting the message into
 a buffer with 6-bytes of pre-padding.
 
+## Wire Type-Length-Value Format
+
+Throughout the protocol, we use a TLV (Type-Length-Value) format to allow for
+the backwards-compatible addition of new fields to existing message types. This
+format describes the TLV encoding specific to wire messages.
+
+A `tlv_record` represents a single field, encoded in the form:
+
+* [`1`: `type`]
+* [`2`: `length`]
+* [`length`: `value`]
+
+A `tlv_stream` is a series of (possibly zero) `tlv_record`s, represented as the
+concatenation of the encoded `tlv_record`s. When used to extend existing
+messages, a `tlv_stream` is typically placed after all currently defined fields.
+
+The `type` is a single-byte identifier for the `tlv_record` and determines how
+the contents of `value` should be decoded. The `type` 255 is reserved for use as
+a sentinel value indicating the end of the `tlv_stream`; the length of `type`
+255 is always zero.
+
```

Comment:
I have question that also applies to @rustyrussell  PR: what is the actual use case for the sentinel/end-of-stream type ?
Since readers must know how to parse T and L (and then skip over L bytes if they don't need or understand T)  I don't really see why it is needed ?

Reply:
you're correct that it's not strictly needed. however, one application is that you can concatenate two `tlv_stream`s with different namespaces. assuming the receiver is asserting the ordering of `type`s, it would otherwise be impossible to distinguish between the stream being not-canonical and the second one starting.

it's also useful in signaling to the receiver that all intended fields were serialized, and that nothing got cutoff/dropped. it could make sense to always include the sentinel value so that this is implicit, but i think maybe that could be done on a case-by-base basis depending on the criticality of the message and its contents

i'm sure @rustyrussell has some more ideas as well :)

Reply:
after further discussion, seems we no longer need sentinel values :) removed in 252ecea1d262c9959159358b7fed38f784bcd96a